# Download Fastq data and uncompress it
gunzip SRR13577846.fastq.gz 

# the number of IDs
cat SRR13577846.fastq | grep '^@' | wc -l
# 117525

# total sequence length 
awk ' /^@/ {sub("length=","",$2); sum+=$2} END {print sum}' SRR13577846.fastq
# 6906121575

# explore the data
awk ' /^@/ {print $1} ' SRR13577846.fastq
awk ' /^@/ {print $2} ' SRR13577846.fastq

# install FastQC
conda create -n FastQC
conda activate FastQC
conda install -c bioconda FastQC=0.11.9

### FastQC                                           
fastqc SRR13577846.fastq
# output is a html page and a zip
# in html output
# If there are no yellow bars displayed in the per base sequence quality plot in the FastQC results, this usually indicates that the overall quality of the sequence is relatively high, with no obvious areas of low quality or low-quality sequences. This is a good sign, indicating that your sequencing data has high quality and is less likely to have low-quality bases or sequences.
# Yellow bars typically indicate a decrease in sequence quality at certain positions, possibly due to adapter sequences, low-quality bases, or other issues. If there are no yellow bars in the per base sequence quality plot, then your data quality does not have any apparent issues, and you can proceed with further data analysis or genome assembly operations.

### install Trimmomatic
conda create -n Trim
conda activate Trim
conda install -c bioconda Trimmomatic=0.39
'
# trim the sequences
java -jar ~/bin/BINP28/Trimmomatic-0.39/trimmomatic-0.39.jar SE ../Raw_data/SRR13577846.fastq.gz SRR.trim.fq.gz AVGQUAL:28 MINLEN:46 CROP:46 SLIDINGWINDOW:8:20
gunzip -c SRR.trim.fq.gz > SRR.trim.fastqc

# fastqc after trimming
fastqc SRR.trim.fastq --threads 6 --outdir
ls
# output files
# SRR.trim.fastq  SRR.trim_fastqc.html  SRR.trim_fastqc.zip  SRR.trim.fq.gz

### Hifiasm
hifiasm -o SRR.asm -t 10 ../Raw_data/SRR13577846.fastq.gz 
ls
# SRR.asm.bp.hap1.p_ctg.gfa        SRR.asm.bp.p_ctg.gfa        SRR.asm.bp.r_utg.gfa
# SRR.asm.bp.hap1.p_ctg.lowQ.bed   SRR.asm.bp.p_ctg.lowQ.bed   SRR.asm.bp.r_utg.lowQ.bed
# SRR.asm.bp.hap1.p_ctg.noseq.gfa  SRR.asm.bp.p_ctg.noseq.gfa  SRR.asm.bp.r_utg.noseq.gfa
# SRR.asm.bp.hap2.p_ctg.gfa        SRR.asm.bp.p_utg.gfa        SRR.asm.ec.bin
# SRR.asm.bp.hap2.p_ctg.lowQ.bed   SRR.asm.bp.p_utg.lowQ.bed   SRR.asm.ovlp.reverse.bin
# SRR.asm.bp.hap2.p_ctg.noseq.gfa  SRR.asm.bp.p_utg.noseq.gfa  SRR.asm.ovlp.source.bin

# the description for the output files:
# SRR.asm.bp.hap1.p_ctg.gfa and SRR.asm.bp.hap2.p_ctg.gfa: These files contain the partially phased contigs representing the two haplotypes in a diploid genome. They may include switch errors and are derived from the primary contigs
# SRR.asm.bp.p_ctg.gfa: This file contains the primary contigs generated by the assembly process. These are the main contigs representing the assembled sequences.
# SRR.asm.bp.r_utg.gfa and SRR.asm.bp.p_utg.gfa: These files contain the resolved and unresolved unitigs, respectively. Unitigs are small DNA sequences that are part of the assembly process.
# SRR.asm.ec.bin: This file contains error correction information generated during the assembly process.
# SRR.asm.ovlp.reverse.bin and SRR.asm.ovlp.source.bin: These files contain overlap information used during the assembly process.
# *.lowQ.bed: These files may contain regions of low quality in the assembled sequences, indicated by their genomic coordinates.
# *.noseq.gfa: These files may contain contigs or unitigs without sequence information, likely representing structural information or connections between contigs.

# here, we need to use SRR.asm.bp.p_ctg.gfa file for later use
# because SRR.asm.bp.p_ctg.gfa contains the primary contigs

# for next step, we need to convert the gfa file into a fasta file
awk '/^S/ {print ">" $2; print $3} ' SRR.asm.bp.p_ctg.gfa > SRR.asm.bp.p_ctg.fasta

# calculate the number of contigs manually
cat SRR.asm.bp.p_ctg.fasta | grep -c '^>'
49
# explore the contigs whether they have non-standard nucleotides
cat SRR.asm.bp.p_ctg.fasta | grep -v \> | tr -d "\n"acgtACGT | wc -c
0

# calculate the GC content for each contig manully
awk '/^>/ {id=$1; total=0;gc=0;next} {for (i=1; i<=length($1);i++) {if(substr($1,i,1)=="G" || substr($1,i,1) == "C") gc++; total=length($1)} gc_count=gc/total; printf("%s %.2f\n", id, gc_count)} ' SRR.asm.bp.p_ctg.fasta 
# >ptg000001l 0.39
# >ptg000002l 0.38
# >ptg000003l 0.38
# >ptg000004l 0.38
# >ptg000005l 0.38
# >ptg000006l 0.39
# >ptg000007l 0.38
# >ptg000008l 0.38
# >ptg000009l 0.39
# >ptg000010l 0.39
# >ptg000011l 0.38
# >ptg000012l 0.40
# >ptg000013l 0.38
# >ptg000014l 0.39
# >ptg000015l 0.39
# >ptg000016l 0.39
# >ptg000017l 0.38
# >ptg000018l 0.16
# >ptg000019l 0.44
# >ptg000020l 0.44
# >ptg000021l 0.45
# >ptg000022l 0.44
# >ptg000023l 0.44
# >ptg000024l 0.44
# >ptg000025l 0.17
# >ptg000026l 0.44
# >ptg000027l 0.44
# >ptg000028l 0.45
# >ptg000029l 0.44
# >ptg000030l 0.44
# >ptg000031l 0.44
# >ptg000032l 0.44
# >ptg000033l 0.44
# >ptg000034l 0.18
# >ptg000035l 0.39
# >ptg000036l 0.44
# >ptg000037l 0.44
# >ptg000038l 0.43
# >ptg000039l 0.15
# >ptg000040l 0.39
# >ptg000041l 0.44
# >ptg000042l 0.44
# >ptg000043l 0.44
# >ptg000044l 0.44
# >ptg000045l 0.44
# >ptg000046l 0.44
# >ptg000047l 0.44
# >ptg000048l 0.44
# >ptg000049l 0.44

# the largest contig
cat SRR.asm.bp.p_ctg.fasta | awk ' {print length($0)}' | sort -n
# 1506376

# total length
cat SRR.asm.bp.p_ctg.fasta | grep -v \> | tr -d '\n' | wc -c
12730920

### Quast
conda create -c Quast
conda activate Quast
conda install -c bioconda Quast

quast.py ../2_Assembly/SRR.asm.bp.p_ctg.fasta 
ls
# basic_stats  icarus_viewers  report.html  report.tex  report.txt             transposed_report.tsv
# icarus.html  quast.log       report.pdf   report.tsv  transposed_report.tex  transposed_report.txt

# Statistics without reference	SRR.asm.bp.p_ctg
# # contigs	49
# # contigs (>= 0 bp)	49
# # contigs (>= 1000 bp)	49
# # contigs (>= 5000 bp)	49
# # contigs (>= 10000 bp)	49
# # contigs (>= 25000 bp)	21
# # contigs (>= 50000 bp)	17
# Largest contig	1506376
# Total length	12730920
# Total length (>= 0 bp)	12730920
# Total length (>= 1000 bp)	12730920
# Total length (>= 5000 bp)	12730920
# Total length (>= 10000 bp)	12730920
# Total length (>= 25000 bp)	12243602
# Total length (>= 50000 bp)	12121907
# N50	809047
# N90	350316
# auN	820916
# L50	6
# L90	15
# GC (%)	38.44
# Mismatches	
# # N's per 100 kbp	0
# # N's	0

# check the version of Quast and Hifiasm
(Quast) [inf-39-2023@bioinf-biol302439 Quast]$ quast --version
# QUAST v5.2.0
(Hifi) [inf-39-2023@bioinf-biol302439 Quast]$ hifiasm --version
# 0.19.8-r603

### Busco
conda create -n Busco
conda activate Busco
conda install -c bioconda busco=5.6.1

# orthology database of Saccharomyces cerevisiae is saccharomycetes_odb10
busco -i ../2_Assembly/SRR.asm.bp.p_ -m genome -l saccharomycetes_odb10 --cpu 10 

# output we need : short_summary.specific.saccharomycetes_odb10.BUSCO_SRR.asm.bp.p_ctg.fasta.txt

# download refernence genome 
# https://www.ncbi.nlm.nih.gov/datasets/taxonomy/4932/

quast.py -r ~/bin/BINP29/Denovo/ncbi_dataset/data/GCF_000146045.2/GCF_000146045.2_R64_genomic.fna ../2_Assembly/SRR.asm.bp.p_ctg.fasta

# Genome statistics	SRR.asm.bp.p_ctg
# Genome fraction (%)	97.42
# Duplication ratio	1.065
# Largest alignment	843303
# Total aligned length	12608885
# NGA50	284119
# LGA50	12
# Misassemblies	
# # misassemblies	119
# Misassembled contigs length	12606913
# Mismatches	
# # mismatches per 100 kbp	205.78
# # indels per 100 kbp	27.64
# # N's per 100 kbp	0
# Statistics without reference	
# # contigs	49
# Largest contig	1506376
# Total length	12730920
# Total length (>= 1000 bp)	12730920
# Total length (>= 10000 bp)	12730920
# Total length (>= 50000 bp)	12121907
# extended part ...

